src/EventListener/WriteListener.php

```diff
@@ -61,7 +61,8 @@ public function onKernelView(GetResponseForControllerResultEvent $event)
 
                 $event->setControllerResult($persistResult ?? $controllerResult);
 
-                if (null !== $this->iriConverter) {
+                // Comparing the class is necessary because the input might not be readable
+                if (null !== $this->iriConverter && \get_class($controllerResult) === \get_class($event->getControllerResult())) {
```

Comment:
Sorry but I'm not sure to understand...

Reply:
> `_api_write_item_iri` should return the IRI of the class pre-persist anyway. It just doesn't make sense if the classes aren't the same anymore.

I don't understand what's the problem. The IRI should not change.

Anyway, can't we just do:

```php
$request->attributes->set('_api_write_item_iri', $this->iriConverter->getIriFromItem($event->getControllerResult()));
```

Reply:
I wanted to do: `$controllerResult = $persistResult ?? $controllerResult` but I think that it breaks the functionality of `_api_write_item_iri` if I understand things correctly.

Reply:
I don't think it'd break anything? The new object returned by the data persister (if any) is what we will serialize. But the IRI should always remain the same.

Reply:
Nope, check the test named `WriteListenerTest::testOnKernelViewWithControllerResultAndPersistWithImmutableResource`

Reply:
@teohhanhui Can you confirm that the test `testOnKernelViewWithControllerResultAndPersistWithImmutableResource` is correct? If I  mutate `$controllerResult` it will fail and, to me, this indicates that I break things so any ideas of what I should do here?